sa: add support for interface suffix for IPv6ll

cspiel1 opened the pull request:
Opened for discussion.

HTTP client already works. Example:
```
echo "/http_get http://[fe80::3a28:d8d8:ddc3:25dd%eth0]:8000/" | nc -N localhost 5555
```

cspiel1 commented:
Still not working:
```
/dial [fe80::3a28:d8d8:ddc3:25dd%eth0]
```

alfredh commented:
this is really great :)

it looks like a simple and non-intrusive solution

alfredh commented on src/sa/sa.c:
```diff
@@ -17,6 +20,30 @@
 #define DEBUG_LEVEL 5
 #include <re_dbg.h>
 
+static int resolve_address(const char *addr, struct sa *sa)
```
should we move this to net.c and make it generic ?

cspiel1 replied:
Good idea!

cspiel1 replied:
`getaddrinfo()` also does a DNS lookup if the format of addr is not an IP address. The timeout can be set in resolve.conf and is by default 5 seconds. The DNS lookup is very fast if the DNS server is reachable, even if the queried address is not reachable. Anyway, we don't want a synchronous DNS lookup.

cspiel1 replied:
Should we add a check in `net_resolve_address()`:
```
@@ -236,6 +236,10 @@ int net_resolve_address(const char *addr, struct sa *sa)
        struct addrinfo *res, *res0;
        int err;
 
+       err = net_inet_pton(addr, sa);
+       if (err)
+               return err;
+
        err = getaddrinfo(addr, NULL, NULL, &res0);
        if (err) {
                DEBUG_WARNING("getaddrinfo(): (%s)\n", gai_strerror(err));
```

For this we would have to move `net_inet_pton()` from internal sa.h to re_sa.h.

cspiel1 replied:
Or we rename the function to `net_resolve_ipv6ll()` and move the simple check which is now in sa.c to `net_resolve_ipv6ll()`:
```
if (!strncmp(addr, "fe80:", 5))
```

alfredh commented on src/sa/sa.c:
```diff
@@ -69,7 +96,11 @@ int sa_set_str(struct sa *sa, const char *addr, uint16_t port)
 		return EINVAL;
 
-	err = net_inet_pton(addr, sa);
+	if (!re_regex(addr, strlen(addr), "%"))
```
perhaps more efficient to do this:

```c
if (strchr(addr, '%'))
```

alfredh commented:
please note that inet_pton and inet_ntop contains 2 backup implementations
that might be buggy.

we should consider to delete the backup implementation, and require
the OS to support those functions.

they are conforming to POSIX.

sreimers commented:
> please note that inet_pton and inet_ntop contains 2 backup implementations
> that might be buggy.

I think we can increase Windows SDK from `_WIN32_WINNT=0x0501` (XP) to `0x600` (Vista), after this inet_pton and inet_ntop should be natively supported. Maybe we should increase to windows 7 (0x0601) right away.

alfredh commented on src/net/net.c:
```diff
@@ -220,3 +220,38 @@ int net_default_gateway_get(int af, struct sa *gw)
+	struct addrinfo *res, *res0;
+	int err;
+
+	err = getaddrinfo(addr, NULL, NULL, &res0);
```
setting ai_flags AI_NUMERICHOST will disable DNS query:

```
                   AI_NUMERICHOST  If the AI_NUMERICHOST bit is set, it indicates that
                                    hostname should be treated as a numeric string
                                    defining an IPv4 or IPv6 address and no name reso-
                                    lution should be attempted.
```

see udp.c and tcp.c

cspiel1 replied:
perfect, thanks!

cspiel1 replied:
It works. Will add a test case to retest that also checks that resolution is disabled.

Should we always call `net_resolve_address()` from `sa_set_str()`?

cspiel1 replied:
I tried this. The `net_inet_pton()` seems to convert IPv6-mapped IPv4 addresses to the associated IPv4 address. And `net_resolve_address()` does not. Is this conversion mandatory?

cspiel1 replied:
Another problem. This test fails:
```
	err = net_resolve_address("6002", &sa);
	TEST_EQUALS(EADDRNOTAVAIL, err);
```
"6002" is not a valid IP address.

Should we add a `re_regex` check for IPv6 and one for IPv4 in `net_resolve_address()`?

alfredh commented on src/net/net.c:
```diff
@@ -220,3 +220,42 @@ int net_default_gateway_get(int af, struct sa *gw)
+
+		break;
+	}
+
```
also need freeaddrinfo here

alfredh commented:
I think we should be a bit careful here and only make the test in the http client:

```c
if (some_condition)
    net_getaddrinfo(sa, addr);
else
    sa_set_str(sa, addr, port);
```

modifying `sa_set_str` is a bit risky.

we should test that the solution works on some platforms, like:

- Linux
- OSX
- Android
- iOS
- Windows
- OpenBSD

cspiel1 commented:
> I think we should be a bit careful here and only make the test in the http client:
> 
> ```c
> if (some_condition)
>     net_getaddrinfo(sa, addr);
> else
>     sa_set_str(sa, addr, port);
> ```
> 
> modifying `sa_set_str` is a bit risky.

Ok, then I remove the commit ed605db. The condition has nothing to do with http, it is for the IPv6ll ifname suffix:
```
if (!strncmp(addr, "fe80:", 5) && strchr(addr, '%'))
```

> 
> we should test that the solution works on some platforms, like:
> 
> * Linux
> * OSX
> * Android
> * iOS
> * Windows
> * OpenBSD

The retest/github workflow should do this. I have only linux. This is the PR for retest: https://github.com/baresip/retest/pull/24

sreimers commented:
In PR #118 I add `sa_pton` so we can use something like this:
```diff
@@ -235,28 +235,18 @@ int net_resolve_address(const char *addr, struct sa *sa)
 {
        struct addrinfo *res, *res0 = NULL;
        struct addrinfo hints;
-       struct pl pl;
        int err;
 
        if (!str_isset(addr) || !sa)
                return EINVAL;
 
-       /* Try IPv6 first */
-       err = re_regex(addr, strlen(addr), "[0-9a-f]+:[0-9a-f]*:[0-9a-f:]+",
-                       &pl, NULL, NULL);
-
-       if (!err && pl.p>addr)
-               return EINVAL;
-
-       /* Then non-IPv6 host */
-       if (err)
-               err = re_regex(addr, strlen(addr),
-                              "[0-9]+.[0-9]+.[0-9]+.[0-9]+",
-                              NULL, NULL, NULL, NULL);
+       err = sa_pton(addr, sa);
+       if (!err)
+               return 0;
 
-       if (err)
+       if (strncmp(addr, "fe80:", 5) || !strchr(addr, '%'))
                return EINVAL;
-
+       
        memset(&hints, 0, sizeof(hints));
        hints.ai_family = AF_UNSPEC;
        hints.ai_flags  = AI_V4MAPPED | AI_ADDRCONFIG | AI_NUMERICHOST;
```
And maybe we should avoid `resolve` in the name, since there is no dns resolution.

cspiel1 commented:
I thought, since resolve is a bad name. Called it net_pton(). But why not doing all in sa_pton()?
The IPv6ll is a special case. Moved the code into sa_pton() as special case. Squashed everything to one commit, cleanup.

sreimers commented:
> I thought, since resolve is a bad name. Called it net_pton(). But why not doing all in sa_pton()?

I like the idea. Looks great. If we found more cases for getaddrinfo we can extend this.

sreimers commented on src/sa/sa.c:
```diff
@@ -12,10 +12,14 @@
 #include <re_fmt.h>
 #include <re_list.h>
 #include <re_sa.h>
+#include <re_net.h>
```
re_net.h seems to be unused now.

cspiel1 replied:
removed the line from the commit

alfredh commented on src/sa/sa.c:
```diff
@@ -86,12 +93,31 @@ int sa_pton(const char *addr, struct sa *sa)
 			sa->u.in6.sin6_family = AF_INET6;
 		}
 	}
+	else if (!strncmp(addr, "fe80:", 5) && strchr(addr, '%')) {
```
I think the previous if statement will be triggered first.
inet_pton on linux or osx handles % without reading the value

cspiel1 replied:
`inet_pton()` is preferred. If it does not work, `getaddrinfo()` is tried. Should I change something?

alfredh commented on src/sa/sa.c:
```diff
@@ -86,12 +93,31 @@ int sa_pton(const char *addr, struct sa *sa)
+			break;
+		}
+
+		freeaddrinfo(res0);
```
I think this block should be a separate function, so that it can be reused

alfredh commented:
we can also consider to drop the check for % and let getaddrinfo convert
all fe80:: addresses for now.

cspiel1 commented:
Ok. I think I know what you mean. Will change the commit again.

alfredh commented:
I think this patch looks good now :)

sreimers commented:
Looks like there is a broken macos test case now, I'll take a closer look later.

sreimers commented:
Interesting we use a invalid ipv6 link local (`fe80::/64` ) address in our tests (and macos getaddrinfo seems to be more strict):
https://github.com/baresip/retest/blob/master/src/sa.c#L121-L122
https://github.com/baresip/retest/blob/master/src/sa.c#L193

Can you replace `fe80:c827:1507:7707:7b75:5489:feb7:2c45` for example with `fe80::215:58ff:fe2d:90ab`

cspiel1 commented:
Great finding. Thanks!

jungle-boogie commented:
Please see https://github.com/baresip/re/issues/124

Maybe more work is done to get this building on OpenBSD i386 platforms?
